task.c

```diff
@@ -0,0 +1,120 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "task.h"
+#include "common.h"
+#include "sockets.h"
+
+/* Tasks are stored in a consecutive chunk of memory, the first
+ * sizeof(task_spec) bytes are arranged acording to the struct
+ * task_spec. Then there is an array of task_args of length
+ * (num_args + num_returns), and then follows the data of
+ * pass-by-value arguments of size args_value_size. The offsets in the
+ * task_arg.val are with respect to the end of the struct, i.e.
+ * with respect to the address &task_spec.ids[0]. */
+
+typedef struct {
+  /* Either ARG_BY_REF or ARG_BY_VAL. */
+  int8_t type;
+  union {
+    unique_id arg_id;
+    struct {
+      /* Offset where the data associated to this arg is located relative
+       * to &task_spec.ids[0]. */
+      ptrdiff_t offset;
+      int64_t length;
+    } val;
+  };
+} task_arg;
+
+struct task_spec_impl {
+  function_id func_id;
+  /* Total number of arguments. */
+  int64_t num_args;
+  /* Position of the argument that is currently constructed. */
+  int64_t arg_cursor;
+  /* Number of return values. */
+  int64_t num_returns;
+  /* Number of bytes the pass-by-value arguments are occupying. */
+  int64_t args_value_size;
+  /* Offset in the argument data. */
+  int64_t args_value_cursor;
+  /* Argument and return IDs. */
+  task_arg ids[0];
+};
+
+task_spec *alloc_task_spec(function_id func_id,
+                           int64_t num_args,
+                           int64_t num_returns,
+                           int64_t args_value_size) {
+  int64_t size = sizeof(task_spec) +
+                 (num_args + num_returns) * sizeof(task_arg) + args_value_size;
+  task_spec *task = malloc(size);
+  memset(task, 0, size);
+  task->func_id = func_id;
+  task->num_args = num_args;
+  task->arg_cursor = 0;
+  task->num_returns = num_returns;
+  task->args_value_size = args_value_size;
```

Comment:
I'd add a member:
task->args_offset = task +sizeof(task_spec);
task->results_offset = task->args_offset + num_args * sizeof(task_arg);

This would make the code easier to read when we index into the arguments or results.